### PR TITLE
Fix documentation for bilinear upsampling and add unit test

### PR DESCRIPTION
**Pull request by vandanavk**
## Description ##
Fixes https://github.com/apache/incubator-mxnet/issues/13078

Future work/alternate approaches:
1. Initialize weight in the backend, so that the user doesn't have to give weights as input.
2. Change backend to call contrib.BilinearResize2D

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Additional documentation
- test

## Comments ##
Ref:
https://github.com/apache/incubator-mxnet/issues/9138
https://stackoverflow.com/questions/47897924/implementing-bilinear-interpolation-with-mxnet-ndarray-upsampling
https://github.com/apache/incubator-mxnet/issues/12970
https://github.com/apache/incubator-mxnet/pull/12983
https://github.com/apache/incubator-mxnet/issues/1412
https://discuss.mxnet.io/t/use-mx-nd-upsampling-with-bilinear-sample-type/2067/2

### CODE REVIEW

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-work-in-progress, Operator]

**Comment by lupesko** [comment]
@apeforest @anirudh2290 can you please help review?

File: python/mxnet/base.py [hunk 1]
```diff
@@ -483,8 +483,6 @@ def build_param_doc(arg_names, arg_types, arg_descs, remove_dup=True):
     for key, type_info, desc in zip(arg_names, arg_types, arg_descs):
         if key in param_keys and remove_dup:
             continue
-        if key == 'num_args':
```
Comment (anirudh2290) [inline python/mxnet/base.py]: num_args is populated for variable length inputs without users having to worry about it. If we remove this the additional num_args field can confuse users.

File: src/operator/nn/upsampling-inl.h [hunk 2]
```diff
@@ -59,24 +59,26 @@ struct UpSamplingParam : public dmlc::Parameter<UpSamplingParam> {
...
     .set_default(0);
     DMLC_DECLARE_FIELD(sample_type)
     .add_enum("nearest", up_enum::kNearest)
     .add_enum("bilinear", up_enum::kBilinear)
     .describe("upsampling method");
+    DMLC_DECLARE_FIELD(num_args).set_default(1)
+    .describe("Number of inputs to be upsampled. For nearest neighbor "
+    "upsampling, this can be 1-N; the size of output will be"
+    "(scale*h_0,scale*w_0) and all other inputs will be upsampled to the"
+    "same size. For bilinear upsampling this must be 2; 1 input and 1 weight.");
```
Comment (anirudh2290) [inline src/operator/nn/upsampling-inl.h]: We can just move this information to the description of data field.

File: tests/python/unittest/test_operator.py [hunk 3]
```diff
@@ -1514,6 +1527,20 @@ def test_nearest_upsampling():
                     check_nearest_upsampling_with_shape(shapes, scale, root_scale)
 
 
+@with_seed()
+def test_bilinear_upsampling():
+    for root_scale in [2,3]:
```
Comment (anirudhacharya) [inline tests/python/unittest/test_operator.py]: you could use `itertools.product` like here - https://github.com/anirudhacharya/incubator-mxnet/blob/master/tests/python/unittest/test_optimizer.py#L1225

File: python/mxnet/base.py (hunk 1 above)
Comment (vandanavk) [reply python/mxnet/base.py]: ok, then maybe I will write about providing 2 inputs for bilinear upsampling in the operator description.

File: src/operator/nn/upsampling-inl.h [hunk 4]
```diff
@@ -59,24 +59,26 @@ struct UpSamplingParam : public dmlc::Parameter<UpSamplingParam> {
...
-    .describe("Input filter. Only used by bilinear sample_type.")
+    .describe("Input filter. Only used by bilinear sample_type. "
+              "Since bilinear upsampling uses deconvolution, num_filters "
+              "is set to the number of channels.")
     .set_default(0);
     DMLC_DECLARE_FIELD(sample_type)
     .add_enum("nearest", up_enum::kNearest)
     .add_enum("bilinear", up_enum::kBilinear)
     .describe("upsampling method");
+    DMLC_DECLARE_FIELD(num_args).set_default(1)
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply src/operator/nn/upsampling-inl.h]: missed changing this back. will do that

File: python/mxnet/base.py (hunk 1 above)
Comment (vandanavk) [reply python/mxnet/base.py]: ok, then i'll just add a note in the description about providing 2 inputs for bilinear sampling

File: src/operator/nn/upsampling-inl.h (hunk 2 above)
Comment (vandanavk) [reply src/operator/nn/upsampling-inl.h]: ok

File: tests/python/unittest/test_operator.py (hunk 3 above)
Comment (vandanavk) [reply tests/python/unittest/test_operator.py]: ok

File: tests/python/unittest/test_operator.py [hunk 5]
```diff
@@ -1491,17 +1491,30 @@ def check_nearest_upsampling_with_shape(shapes, scale, root_scale):
         assert_allclose(arr[name].asnumpy()*root_scale**2*scale**(2*k), arr_grad[name].asnumpy(), rtol=1e-4)
 
 
-def check_bilinear_upsampling_with_shape(shapes, scale, root_scale):
-    arr = {'arg_%d'%i: mx.random.uniform(-10.0, 10.0, shape, ctx=mx.cpu()).copyto(default_context()) for i, shape in zip(range(len(shapes)), shapes)}
-    arr_grad = {'arg_%d'%i: mx.nd.zeros(shape) for i, shape in zip(range(len(shapes)), shapes)}
-
-    up = mx.sym.UpSampling(*[mx.sym.Variable('arg_%d'%i) for i in range(len(shapes))], sample_type='bilinear', scale=root_scale)
+def check_bilinear_upsampling_with_shape(data_shape, weight_shape, scale, root_scale, num_filter):
```
Comment (apeforest) [inline tests/python/unittest/test_operator.py]: Why did you change the unit test if your PR is to fix documentation?

Comment (vandanavk) [reply tests/python/unittest/test_operator.py]: @apeforest the unit test was not being called previously. And based on inputs from the issues in Ref section of the PR description, many users are confused about how to use the Upsampling operator with the bilinear upsampling option. So i thought I could add/modify tests here as well - which would show how to use the operator. 
I have updated the PR title - should I open a separate PR instead?

Note: This PR is WIP as an assert has to be added at the end of the test. Based on an offline discussion with @anirudh2290 yesterday, the assert will be similar to what was done for test_deconvolution()

**Comment by vandanavk** [comment]
@mxnet-label-bot update [Operator, pr-awaiting-review]

**Comment by piyushghai** [comment]
@apeforest @anirudhacharya Are you comments on the PR addressed now ? Can this PR be taken forward now ?

**Comment by Roshrini** [comment]
@anirudh2290 @apeforest Can you take another look at this PR?
